sharding/shard.go

```diff
@@ -0,0 +1,262 @@
+	if err != nil {
+		return nil, fmt.Errorf("error while getting canonical header hash: %v", err)
+	}
+
+	collation, err := s.CollationByHash(h)
+	if err != nil {
+		return nil, fmt.Errorf("no canonical collation found for hash: %v", err)
+	}
+	return collation, nil
+}
+
+// BodyByChunkRoot fetches a collation body.
+func (s *Shard) BodyByChunkRoot(chunkRoot *common.Hash) ([]byte, error) {
+	body, err := s.shardDB.Get(*chunkRoot)
+	if err != nil {
+		return nil, err
+	}
+	if body == nil {
+		return nil, fmt.Errorf("no corresponding body with chunk root found: %v", chunkRoot.String())
+	}
+	return *body, nil
+}
+
+// CheckAvailability is used by notaries to confirm a header's data availability.
+func (s *Shard) CheckAvailability(header *CollationHeader) (bool, error) {
+	key := dataAvailabilityLookupKey(header.ChunkRoot())
+	val, err := s.shardDB.Get(key)
+	if err != nil {
+		return false, err
+	}
+	if val == nil {
+		return false, fmt.Errorf("availability not set for header")
+	}
+	availability := *val
+	// availability is a byte array of length 1.
+	return availability[0] != 0, nil
+}
+
+// SetAvailability saves the availability of the chunk root in the shardDB.
+func (s *Shard) SetAvailability(chunkRoot *common.Hash, availability bool) error {
+	key := dataAvailabilityLookupKey(chunkRoot)
+	var encoded []byte
+	if availability {
+		encoded = []byte{1}
+	} else {
+		encoded = []byte{0}
+	}
+	if err := s.shardDB.Put(key, encoded); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SaveHeader adds the collation header to shardDB.
+func (s *Shard) SaveHeader(header *CollationHeader) error {
+	// checks if the header has a chunk root set before saving.
+	if header.ChunkRoot() == nil {
+		return fmt.Errorf("header needs to have a chunk root set before saving")
+	}
+
+	encoded, err := header.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+
+	// uses the hash of the header as the key.
+	if err := s.shardDB.Put(header.Hash(), encoded); err != nil {
+		return fmt.Errorf("cannot update shardDB: %v", err)
+	}
+	return nil
+}
+
+// SaveBody adds the collation body to the shardDB and sets availability.
+func (s *Shard) SaveBody(body []byte) error {
+	// TODO: check if body is empty and throw error.
+	// TODO: dependent on blob serialization.
+	// chunkRoot := getChunkRoot(body) using the blob algorithm utils.
+	// right now we will just take the raw keccak256 of the body until #92 is merged.
+	chunkRoot := common.BytesToHash(body)
+
+	if err := s.shardDB.Put(chunkRoot, body); err != nil {
+		return err
+	}
+	s.SetAvailability(&chunkRoot, true)
+	return nil
+}
+
+// SaveCollation adds the collation's header and body to shardDB.
+func (s *Shard) SaveCollation(collation *Collation) error {
+	if err := s.ValidateShardID(collation.Header()); err != nil {
+		return err
+	}
+	if err := s.SaveHeader(collation.Header()); err != nil {
+		return err
+	}
+	if err := s.SaveBody(collation.Body()); err != nil {
```

Comment:
Just return this one since you aren't doing anything with the error.

I'm surprised the linter isn't blowing up on this. Seems like something a linter would recommend. I am a human linter now